File: 173_BST_Iterator.py
# Leetcode problem link : https://leetcode.com/problems/binary-search-tree-iterator/
# Time Complexity : O(n)
#  Space Complexity : O(1) 
#  Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
'''
        Basic approach : O(n) => Create an array with inorder traversal of the array and keep track of current index
    
        Optimized approach: O(maxdepth) space => The smallest element in BST is the lftmost one so we can push all the elements till the leftmost one in stack and keep popping every element and check for its right subtree and push its left branch elements. Repeat.
    '''  

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: TreeNode):
        self.stack = []
        self._dfs(root)
    
    def _dfs(self,root):
        while root:
            self.stack.append(root)
            root = root.left

    def next(self) -> int:
        """
        @return the next smallest number
        """
        if not self.hasNext:
            return -1
        
        nextVal = self.stack.pop()
        if (nextVal.right):
            self._dfs(nextVal.right)
        
        return nextVal.val
        

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return len(self.stack) != 0
# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: 173-BSTIterator.java
/**LC-173 Design Problem. 
 * 
 * Time Complexity : O(1) 
 * Space Complexity : O(H) H: Height of the tree
 * Did this code successfully run on Leetcode : Yes
 * Any problem you faced while coding this : No
 *
 Algorithm: Main idea, keep the next element on the top of the stack always
 1. In the constructor itself, put each node from root to the left most bottom element, which is the smallest elememt
 2. When next() function is called, pop the top value. Pop happens when the top element does not has left child
 3. Now do the same left recursion to the right side of the popped element (if the popped element has a right side)
 
 While finding the space and time complexity, we don't bother about the constructor since it is not user centric.
 and also called only once. User is not going to use the constructor, user will use next() and hasNext()

 hasNext() -> O(1)
 next() Average O(1) ...when we removed the 1 it didn't had right child, at that time the time complexity was O(1)
 */

 /**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    Stack<TreeNode> st;

    public BSTIterator(TreeNode root) {
        
        st = new Stack<>();
        dfs(root); 
    }
    
    /** @return the next smallest number */
    public int next() {
        
        TreeNode returnedNode = st.pop();  //O(1)
        if(returnedNode.right != null){
            dfs(returnedNode.right);
        }
        return returnedNode.val;
    }
    
    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return !st.isEmpty();    //O(1)
    }
    
    private void dfs(TreeNode root){
        
        while(root != null){
            
            st.push(root);
            root = root.left;
        }
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */